

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Input files &mdash; mrbait 1.1.2 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Usage options" href="usage.html" />
    <link rel="prev" title="Getting Started" href="installation.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> mrbait
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="introduction.html">Introduction</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="pipeline.html">Pipeline Description</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html">Getting Started</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Input files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#assembled-genomes">Assembled genomes</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#annotating-genomes-with-vcf">Annotating genomes with VCF</a></li>
<li class="toctree-l4"><a class="reference internal" href="#annotating-genomes-with-gff">Annotating genomes with GFF</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#multiple-genome-alignments">Multiple genome alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reduced-representation-data">Reduced representation data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="usage.html">Usage options</a></li>
<li class="toctree-l2"><a class="reference internal" href="output.html">Output Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="benchmarking.html">Benchmarking and Hardware Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="acknowledgements.html">Acknowledgements</a></li>
<li class="toctree-l2"><a class="reference internal" href="references.html">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pipeline.html">Pipeline Description</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Getting Started</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Input files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#assembled-genomes">Assembled genomes</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#annotating-genomes-with-vcf">Annotating genomes with VCF</a></li>
<li class="toctree-l3"><a class="reference internal" href="#annotating-genomes-with-gff">Annotating genomes with GFF</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#multiple-genome-alignments">Multiple genome alignments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reduced-representation-data">Reduced representation data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage options</a></li>
<li class="toctree-l1"><a class="reference internal" href="output.html">Output Files</a></li>
<li class="toctree-l1"><a class="reference internal" href="benchmarking.html">Benchmarking and Hardware Requirements</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgements.html">Acknowledgements</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">mrbait</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="introduction.html">Introduction</a> &raquo;</li>
        
      <li>Input files</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/files.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="input-files">
<span id="files"></span><h1>Input files<a class="headerlink" href="#input-files" title="Permalink to this headline">¶</a></h1>
<p>This section describes the input file types accepted by MrBait.</p>
<div class="section" id="assembled-genomes">
<h2>Assembled genomes<a class="headerlink" href="#assembled-genomes" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://github.com/tkchafin/mrbait">mrbait</a> only accepts genome assemblies formatted as FASTA. These can represent
contigs, scaffolds, or entire chromosomes. According to the FASTA specifications,
a sequence should begin with a header line, or short description (indicated by the
“&gt;” symbol), followed by a second line containing sequence data. It does
not matter if the following lines are interleaved or on a single line, and any blank
lines in the file will be ignored, as will any leading or trailing whitespace.</p>
<p>An example FASTA-formatted sequence is given below.</p>
<div class="highlight-none notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span>&gt;chr1.scaffold1
ATAGCTCGGCTACGTGATCGCGTGCTC-ATGCTAGCGCTNNNNNNNNATGATTGCTTTT
TGTGTGTGCAAGCACTGCCGRGCTACGCGCTACTGCCRCCTAGTATGTGTGGCCGCTAC
TAGTCCGCGCTAGCTtTtagatctcgtggcgccgcgcgcgtcgcacgatcgtacgcgcc
&gt;chr1.scaffold2
ATCGTGCTGCGGCGCTGCCTCAGC…
…
…
…
</pre></div>
</td></tr></table></div>
<div class="section" id="annotating-genomes-with-vcf">
<h3>Annotating genomes with VCF<a class="headerlink" href="#annotating-genomes-with-vcf" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="https://github.com/tkchafin/mrbait">mrbait</a> also supports supplementing genomic sequences with coordinate-reference SNP
data (e.g. obtained from population-level sequencing) using the <a class="reference external" href="http://samtools.github.io/hts-specs/">Variant Call Format</a>:</p>
<div class="highlight-none notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10</pre></div></td><td class="code"><div class="highlight"><pre><span></span>##fileformat=VCFv4.2
##FORMAT=&lt;ID=GT,Number=1,Type=Integer,Description=&quot;Genotype&quot;&gt;
##FORMAT=&lt;ID=GP,Number=G,Type=Float,Description=&quot;Genotype Probabilities&quot;&gt;
##FORMAT=&lt;ID=PL,Number=G,Type=Float,Description=&quot;Phred-scaled Genotype Likelihoods&quot;&gt;
#CHROM       POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  SAMP001 SAMP002
chr1.scaffold1       48      rs11449 G       A       .       PASS    .       GT      0/0     0/1
chr1.scaffold1       47      rs11449 T       A       .       PASS    .       GT      0/0     0/1
chr1.scaffold2       1 rs84825 A     T       .       PASS    .       GT:GP   0/1:.   0/1:0.03,0.97,0
…
…
</pre></div>
</td></tr></table></div>
<p>It is important to note that the VCF format can communicate much more information
than <a class="reference external" href="https://github.com/tkchafin/mrbait">mrbait</a> will utilize. The CHROM and POS columns will be parsed to locate the
reference position for each SNP, and the REF and ALT columns to write a new consensus
base at that position using IUPAC ambiguity codes (e.g. C/T = Y). More functionality
will be added in future versions of <a class="reference external" href="https://github.com/tkchafin/mrbait">mrbait</a>.</p>
<p>It is highly recommended you add variant data if it is available, as it will be used
both for finding adequately conserved regions for bait design, as well as for filtering
target regions for those which capture flanking SNPs.</p>
</div>
<div class="section" id="annotating-genomes-with-gff">
<h3>Annotating genomes with GFF<a class="headerlink" href="#annotating-genomes-with-gff" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="https://github.com/tkchafin/mrbait">mrbait</a> can also make use of genomic features provided using the Generic Feature Format (GFF),
independently or in addition to any variant data provided via VCF. mrbait assumes that input
GFF files follow the version 3 <cite>GFF specification &lt;https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md&gt;</cite>_:</p>
<div class="highlight-none notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6</pre></div></td><td class="code"><div class="highlight"><pre><span></span>##gff-version 3
chr1.scaffold1       .       gene    10      180     .       +       .       ID=gene0001;Alias=targets
chr1.scaffold1       .       mRNA    20      180     .       +       .       ID=mrna0001;Parent=gene0001
chr1.scaffold1       .       exon    10      128     .       +       .       ID=tfbs00001;Parent=gene0001
…
…
</pre></div>
</td></tr></table></div>
<p>Columns should be separated by tabs and defined according to the GFF3 standard (e.g.
column 1 contains the sequence ID). mrbait will use the sequence ID (column 1) to map
coordinates in GFF columns 4 and 5 to the reference provided in your FASTA file, thus
these identifiers must be identical. mrbait will also categorize features internally by
the type (e.g. “exon”) given in column 3, and by any alias assigned in the attributes
column (column 9). All other columns are ignored. You can use either type or alias to
tell <a class="reference external" href="https://github.com/tkchafin/mrbait">mrbait</a> to target those features for bait design.</p>
<p>If you are not targeting all of a single type (e.g. CDS, or exon), you can either pre-filter
your GFF file prior to loading, or you can annotate features of interest using the Alias
attribute.</p>
</div>
</div>
<div class="section" id="multiple-genome-alignments">
<h2>Multiple genome alignments<a class="headerlink" href="#multiple-genome-alignments" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://github.com/tkchafin/mrbait">mrbait</a> reads two different input file types for multiple genome alignments. These can
be provided using the Multiple Alignment Format (<a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format9.3">MAF</a>), or the eXtended Multi-FastA (<a class="reference external" href="https://asap.genetics.wisc.edu/software/mauve/mauve-user-guide/mauve-output-file-formats.php">XMFA</a>) formats.</p>
<p>The MAF format is output by several multiple alignment programs, including <a class="reference external" href="https://mafft.cbrc.jp/alignment/software/">MAFFT</a>
and <a class="reference external" href="http://mugsy.sourceforge.net/">Mugsy</a>, and take the following general form:</p>
<div class="highlight-none notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16</pre></div></td><td class="code"><div class="highlight"><pre><span></span>##maf version=1 scoring=tba.v8
# tba.v8 (((human chimp) baboon) (mouse rat))
# multiz.v7
# maf_project.v5 _tba_right.maf3 mouse _tba_C
# single_cov2.v4 single_cov2 /dev/stdin

a score=5062.0
s hg16.chr7    27699739 6 + 158545518 RAAAGAGATGCTAAGCCAATGAGTTGATGTCTCTCAATGTGTG
s panTro1.chr6 28862317 6 + 161576975 RAAAGAGATGCTAAGCCAATGAGTTGTTGTCTCTCAATGTGTG
s baboon         241163 6 +   4622798 TAAAGAGATGCTAAGCCAATGAGTTGTTGTCTCTRAATGTGTG
s mm4.chr6     53303881 6 + 151104725 TAAAGAGATGCTAAGCCAATGAGTTGTTGTCGCTCAATGTGTG
s rn3.chr4     81444246 6 + 187371129 taaggaGATGCTAAGCCAATGAGTTGTTGTCGCTCAATGTGTG

…
…
…
</pre></div>
</td></tr></table></div>
<p>Comment lines (starting with “#”) are ignored by <a class="reference external" href="https://github.com/tkchafin/mrbait">mrbait</a>. Alignment blocks (considered
by <a class="reference external" href="https://github.com/tkchafin/mrbait">mrbait</a> to each represent different loci) are started with “a”, followed by sequence
lines starting with “s”. Source, strand, and coordinate positions are not informative for
<a class="reference external" href="https://github.com/tkchafin/mrbait">mrbait</a>, nor are lines starting with other letters (which can be used in the MAF
format to communicate additional information about the preceding sequence, such as
quality scores).</p>
<p>The eXtended Multi-FastA (XMFA) format output by the multiple-genome aligner
MAUVE (which outputs it as “.alignment”) is an extension of the standard FASTA format
to allow alignment blocks from many different loci, with header lines representing
identifiers for the aligned sequence, and start-end coordinates representing the alignment
block location within the genome, followed by the sequence:</p>
<div class="highlight-none notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="highlight"><pre><span></span>&gt;1:1-230 +
ATAGC-NAATC--GC…
&gt;2:210-440 -
ATTGGCCAATCCCC…
&gt;3:3-230 +
TTA-CCAAGC--GC…
=
…
…
</pre></div>
</td></tr></table></div>
<p>Alignment blocks are delimited by the “=” symbol. All alignment blocks are assumed
by <a class="reference external" href="https://github.com/tkchafin/mrbait">mrbait</a> to represent separate, discontinuous loci. Note that because
no individual ‘alignment block’ in the .xmfa file is guaranteed to contain the same
genome representatives, no reference coordinates are saved by mrbait. This means
that additional annotation via GFF or VCF cannot be added to whole-genome alignments
provided in .xmfa format.</p>
</div>
<div class="section" id="reduced-representation-data">
<h2>Reduced representation data<a class="headerlink" href="#reduced-representation-data" title="Permalink to this headline">¶</a></h2>
<p>Alignments from reduced-representation methods such as restriction-site associate
DNA sequencing methods (RADseq) can be input using the MAF or XMFA formats, or using
the “.loci” format output by the RADseq assembly pipeline <a class="reference external" href="https://github.com/dereneaton/pyrad">pyrad</a> or
its successor <a class="reference external" href="https://github.com/dereneaton/ipyrad">ipyrad</a>. This format shows individual
loci delimited by a line starting with “//” which features additional annotation of
variants and parsimony-informative sites:</p>
<div class="highlight-none notranslate"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6
7
8</pre></div></td><td class="code"><div class="highlight"><pre><span></span>&gt;PopA001     GTGTGATAGTAGTGATGTATTTTATAATATATATTATCGGATAT……
&gt;PopA002     GTGTGARAGTAGTGATGTATTTTATAATATATATTATCGGATAT……
&gt;PopB001     GTGTGACAGTAGTGATGTATTTTATAATATATATTATCGGATAT……
&gt;PopB002     GAGTGATAGTAGTGATGTATTTTATAATATATATTATCGGATAT……
//            *      *                                        |1
…
…
…
</pre></div>
</td></tr></table></div>
<p>mrbait ignores annotation information (since it parses variants anyways to generate
a consensus sequence), and only uses the “//” delimiter to distinguish between alignment
blocks. Creating a .loci file from other formats can be accomplished relatively easily.
For example, a series of separate alignments (each as .fasta), could be converted to the
<strong>.loci</strong> format using the following <strong>bash</strong> command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> file in <span class="sb">`</span>ls example*.fasta<span class="sb">`</span><span class="p">;</span> <span class="k">do</span>
  awk <span class="s1">&#39;BEGIN{ORS=&quot;&quot;}$1~/^\&gt;/{print $01&quot;\t&quot;;next}{print $0&quot;\n&quot;}&#39;</span> <span class="nv">$file</span>
  &gt;&gt; example.loci<span class="p">;</span>
  <span class="nb">echo</span> <span class="s2">&quot;//&quot;</span> &gt;&gt; example.loci<span class="p">;</span>
<span class="k">done</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="usage.html" class="btn btn-neutral float-right" title="Usage options" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="installation.html" class="btn btn-neutral" title="Getting Started" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Tyler K. Chafin.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.1.2',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>